File: 고득점 kit/완전탐색/전력망을 둘로 나누기/전력망을 둘로 나누기.swift
import Foundation

func solution(_ n:Int, _ wires:[[Int]]) -> Int {
    var answer: Int = Int.max
    // 인접행렬
    var adj: [[Int]] = Array(repeating: Array(repeating: 0,count:n+1 ), count: n+1)
    
    // 인접행렬 표시
    for i in 0..<wires.count {
        adj[wires[i][0]][wires[i][1]] = 1
        adj[wires[i][1]][wires[i][0]] = 1
    }
    
    // BFS
    func BFS(_ start: Int) -> Int {
        // 방문리스트
        var visited:[Bool] = Array(repeating: false, count: n+1)
        visited[start] = true
        
        var que = [start]
        var ret = 1 // 현재 스타트 포함
        
        // 큐가 비어질때까지
        while que.count > 0 {
            let cur = que.removeFirst()
            
            for i in adj[cur].enumerated() {
                // 연결되어 있고 미방문이면
                if i.1 == 1 && visited[i.0] == false {
                    visited[i.0] = true
                    que.append(i.0)
                    ret += 1
                }
            }
        }
        return ret
    }
    
    // 와이어를 하나씩 끈고 DFS 수행하며 최선의 정답 -> 최솟값 필요
    for i in 0..<wires.count {
        // 끊음
        adj[wires[i][0]][wires[i][1]] = 0
        adj[wires[i][1]][wires[i][0]] = 0
        
        let ret1 = BFS(wires[i][0])
        let ret2 = BFS(wires[i][1])
        
        // 최솟값 체크
        answer = min(answer, abs(ret1-ret2))
        // 원상복구
        adj[wires[i][0]][wires[i][1]] = 1
        adj[wires[i][1]][wires[i][0]] = 1
    }
    return answer
}

File: 고득점 kit/완전탐색/전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <cstring>

using namespace std;

int board[101][101];
int visited[101];

int abs(int a) {
    return a < 0 ? -a : a;
}

int min(int a, int b) {
    return a < b ? a : b;
}

void DFS(int n, int cur, int &cnt) {
    visited[cur] = 1;
    cnt++;
    
    for (int i = 1; i <= n; ++i) {
        if (visited[i] == 1)
            continue;
        
        if (board[cur][i] == 0)
            continue;
        
        DFS(n, i, cnt);
    }
}

int solution(int n, vector<vector<int>> wires) {
    int answer = n; // 최솟값을 구해야 하므로
    
    for (auto& w : wires) {
        board[w[0]][w[1]] = 1;
        board[w[1]][w[0]] = 1;
    }
    
    for (auto& w : wires) {
        board[w[0]][w[1]] = 0;
        board[w[1]][w[0]] = 0;
        memset(visited, 0, sizeof(visited));
        
        int cnt1 = 0;
        int cnt2 = 0;
        DFS(n, w[0], cnt1);
        DFS(n, w[1], cnt2);
        answer = min(answer, abs(cnt1 - cnt2));
        
        board[w[0]][w[1]] = 1;
        board[w[1]][w[0]] = 1;
    }
    
    return answer;
}